road/scatter.py

```
@@ -0,0 +1,236 @@
+import os
+import json
+import yaml
+import pandas as pd
+from sklearn.linear_model import LinearRegression
+
+
+def compute_and_save_errors(
+        est_df,
+        obs_df,
+        chosen_timeperiod,
+        combined_df_cols,
+        classification_col):
+    # Compute errors
+    times = ['Daily', 'AM', 'MD', 'PM', 'EV', 'EA']
+    error = {col: est_df[col] - obs_df[col] for col in times}
+    error_df = pd.DataFrame(error)
+    error_squared_df = error_df.pow(2)
+    error_percent_df = error_df.divide(obs_df[times]).fillna(0)
+    error_percent_df = error_percent_df.apply(lambda x: x * 100, axis=1)
+
+    # Combine dataframes
+    combined_df = pd.concat(
+        [est_df, obs_df[times], error_df, error_squared_df, error_percent_df], axis=1)
+    Select_time_period_df = combined_df.filter(like=chosen_timeperiod)
+    specific_columns_combined_df = combined_df[combined_df_cols]
+    Select_time_period_loc_df = pd.concat(
+        [specific_columns_combined_df, Select_time_period_df], axis=1)
+
+    # Set column names
+    calculation_cols = [
+        'Estimated Volume',
+        'Observed Volume',
+        'Errors',
+        'Squared Errors',
+        'Percent Errors']
+    Select_time_period_loc_df.columns = combined_df_cols + calculation_cols
+
+    # Drop duplicates

```

hmm do you know why there would be duplicates created? and is the drop_duplicates(..., keep="first") safe? Like are we throwing away useful information? Do all duplicates always have the same value?

---

they are duplications in the observed data. dont know the exact reason. but you are right, we cant know guarantee if useful information will be removed next time. so we better clean the observed date entirely before the validation

---

@DaweiPang32 if this is in the observed data (sounds like it is), maybe we should raise an Error and crash here? Seems like a RuntimeError would be the best here? (I just looked briefly at https://docs.python.org/3/library/exceptions.html#exception-hierarchy but let me know if you have other thoughts)

---

i think a runtime error is good, I can add it to the validation_road when first loading the obs